File: src/kicad_xyce_plugin/simulation_parameters/noise_simulation_parameters.py
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path

from ..netlist_parser import NetlistTopology
from .measure_parameters import MeasureEntry
from .print_parameters import PrintParameters


def _parse_output_node(token: str) -> tuple[str, str]:
    # match V(node) or V(node,ref)
    m = re.fullmatch(r"[Vv]\(([^,)]+)(?:,([^)]+))?\)", token)
    if m:
        return m.group(1).strip(), (m.group(2) or "").strip()
    # fallback — return the token itself as the output node
    return token, ""


def _parse_device_noise_operator(variable: str) -> DeviceNoiseOperator | None:
    # match DNI(device) or DNO(device,source)
    m = re.fullmatch(r"D(NI|NO)\(([^,)]+)(?:,([^)]+))?\)", variable.upper())
    if m:
        # extract operator type, device name, and optional noise source
        operator_type = f"D{m.group(1)}"
        device_name = m.group(2).strip()
        noise_source = (m.group(3) or "").strip()
        # return parsed operator
        return DeviceNoiseOperator(device_name=device_name, operator_type=operator_type, noise_source=noise_source)
    # return none for non-matching variables
    return None


def _format_device_noise_operator(operator: DeviceNoiseOperator) -> str:
    """Format DeviceNoiseOperator as DNI(device) or DNO(device,source)."""
    # check if noise source is present
    if operator.noise_source:
        # return two-parameter form
        return f"{operator.operator_type}({operator.device_name},{operator.noise_source})"
    # return single-parameter form
    return f"{operator.operator_type}({operator.device_name})"


@dataclass(frozen=True)
class DeviceNoiseOperator:
    device_name: str
    operator_type: str
    noise_source: str = ""


@dataclass(frozen=True)
class NoiseSimulationParameters:

    output_node: str
    ref_node: str = ""
    source_name: str = ""
    sweep_mode: str = "LIN"
    points: str = ""
    start: str = ""
    end: str = ""
    data_table_name: str = ""
    replace_ground: bool = True
    print_parameters: PrintParameters | None = None
    device_noise_operators: tuple[DeviceNoiseOperator, ...] = ()
    measure_parameters: tuple[MeasureEntry, ...] = ()

    @classmethod
    def from_xyce_directives(cls, directives: list[str]) -> "NoiseSimulationParameters" | None:
        # init defaults
        output_node = ""
        ref_node = ""
        source_name = ""
        sweep_mode = "LIN"
        points = ""
        start = ""
        end = ""
        data_table_name = ""
        replace_ground = True
        print_parameters = None
        device_noise_operators: list[DeviceNoiseOperator] = []
        # init measure parameters
        measure_parameters: list[MeasureEntry] = []
        # flag indicating whether a valid directive was found
        found = False
        # parse directives
        for directive in directives:
            # tokenize the directive
            tokens = directive.split()
            # skip empty directives
            if not tokens:
                continue
            cmd = tokens[0].upper()
            # parse print directives and retain noise-specific output config
            if cmd == ".PRINT":
                # parse the print statement from the directive
                print_statement = PrintParameters.from_xyce_statement(directive)
                # retain noise print parameters when found
                if print_statement and print_statement.print_type == "NOISE":
                    # store the parsed print parameters
                    print_parameters = print_statement
                    # extract device noise operators from output variables
                    for variable in print_statement.output_variables:
                        # attempt to parse as device noise operator
                        operator = _parse_device_noise_operator(variable)
                        if operator:
                            # add to list if parsing succeeded
                            device_noise_operators.append(operator)
                    # next
                    continue
            # handle preprocess replaceground
            if cmd == ".PREPROCESS" and len(tokens) > 2 and tokens[1].upper() == "REPLACEGROUND":
                # set replace_ground based on the third token
                replace_ground = tokens[2].upper() == "TRUE"
                # next
                continue
            # parse measure directives
            if cmd in (".MEASURE", ".MEAS"):
                # parse the measure statement from the directive
                measure_statement = MeasureEntry.from_xyce_statement(directive)
                # retain measure parameters when found and analysis type matches
                if measure_statement and measure_statement.analysis_type in ("NOISE", "NOISE_CONT"):
                    # append the parsed measure parameters
                    measure_parameters.append(measure_statement)
                # next
                continue
            # skip non-NOISE directives
            if cmd != ".NOISE":
                continue
            # flag indicating a valid NOISE directive was found
            found = True
            if len(tokens) < 3:
                continue
            # parse output node: tokens[1] is V(out) or V(out,ref)
            output_node, ref_node = _parse_output_node(tokens[1])
            # parse source name: tokens[2]
            source_name = tokens[2]
            if len(tokens) < 4:
                continue
            sweep_token = tokens[3].upper()
            # handle DATA sweep: .NOISE V(...) SRC DATA=<tablename>
            if sweep_token.startswith("DATA="):
                sweep_mode = "DATA"
                data_table_name = tokens[3].split("=", 1)[1]
                continue
            # detect log sweep type
            if sweep_token in ("DEC", "OCT"):
                sweep_mode = sweep_token
                if len(tokens) >= 7:
                    points = tokens[4]
                    start = tokens[5]
                    end = tokens[6]
                continue
            # linear sweep
            sweep_mode = "LIN"
            if sweep_token == "LIN":
                if len(tokens) >= 7:
                    points = tokens[4]
                    start = tokens[5]
                    end = tokens[6]
            else:
                # implicit LIN (no keyword)
                if len(tokens) >= 6:
                    points = tokens[3]
                    start = tokens[4]
                    end = tokens[5]
        # return instance if a valid directive was found
        return cls(output_node=output_node, ref_node=ref_node, source_name=source_name, sweep_mode=sweep_mode, points=points, start=start, end=end, data_table_name=data_table_name, replace_ground=replace_ground, print_parameters=print_parameters, device_noise_operators=tuple(device_noise_operators), measure_parameters=tuple(measure_parameters)) if found else None

    def to_xyce_directives(self, topology: NetlistTopology | None = None) -> list[str]:
        # prepend replaceground preprocessing when enabled
        preprocess = [".PREPROCESS REPLACEGROUND TRUE"] if self.replace_ground else []
        # build the v(out[,ref]) token
        if self.ref_node:
            out_token = f"V({self.output_node},{self.ref_node})"
        else:
            out_token = f"V({self.output_node})"
        # build the sweep portion
        if self.sweep_mode == "DATA":
            sweep = f"DATA={self.data_table_name}"
        elif self.sweep_mode in ("DEC", "OCT"):
            sweep = f"{self.sweep_mode} {self.points} {self.start} {self.end}"
        else:
            sweep = f"LIN {self.points} {self.start} {self.end}"
        # build the noise directive line
        lines = [f".NOISE {out_token} {self.source_name} {sweep}"]
        # append noise print directive when configured
        if self.print_parameters and self.print_parameters.print_type == "NOISE":
            # start with existing output variables
            output_vars = list(self.print_parameters.output_variables)
            # add device noise operators in formatted form
            for operator in self.device_noise_operators:
                # format operator as string
                formatted = _format_device_noise_operator(operator)
                # add if not already present
                if formatted not in output_vars:
                    output_vars.append(formatted)
            # only emit print directive if there are output variables
            if output_vars:
                # create updated print parameters with merged output variables
                updated_print = PrintParameters(print_type="NOISE", print_format=self.print_parameters.print_format, print_file=self.print_parameters.print_file, output_variables=tuple(output_vars), extra_options=self.print_parameters.extra_options)
                # append the updated print statement
                lines.append(updated_print.to_xyce_statement())
        # append measure directives when configured
        for measure in self.measure_parameters:
            lines.append(measure.to_xyce_statement())
        # return the full directive list
        return preprocess + lines

    def raw_output_file_path(self, working_directory: Path, netlist_file_path: Path) -> Path | None:
        # check raw format is selected in print parameters and a print file is specified
        if self.print_parameters is None or self.print_parameters.print_format != "RAW":
            return None
        # return the output file path specified in the print parameters if available
        if self.print_parameters.print_file:
            return working_directory / self.print_parameters.print_file
        # otherwise, create raw file from netlist file, adding the ".raw" suffix
        return netlist_file_path.with_suffix(netlist_file_path.suffix + ".raw")

File: src/simulation/noise_simulation_parameters.cpp
#include <cctype>
#include <optional>
#include <string>
#include <vector>

#include "../core/util.h"
#include "noise_simulation_parameters.h"

DeviceNoiseOperator::DeviceNoiseOperator(std::string type, std::string node, std::string source) :
    type(std::move(type)), node(std::move(node)), source(std::move(source)) {}

bool DeviceNoiseOperator::operator==(const DeviceNoiseOperator& other) const {
    // compare all fields for equality
    return type == other.type && node == other.node && source == other.source;
}

NoiseSimulationParameters::NoiseSimulationParameters(std::string output_node, std::string ref_node, std::string source_name, std::string start_freq_value, std::string end_freq_value, std::string num_points_value, std::string sweep_type, std::vector<DeviceNoiseOperator> device_noise_operators, std::string data_table_name, std::optional<PrintParameters> print_parameters) :
    output_node(std::move(output_node)), ref_node(std::move(ref_node)), source_name(std::move(source_name)), start_freq_value(std::move(start_freq_value)), end_freq_value(std::move(end_freq_value)), num_points_value(std::move(num_points_value)), sweep_type(std::move(sweep_type)), device_noise_operators(std::move(device_noise_operators)), data_table_name(std::move(data_table_name)), print_parameters(std::move(print_parameters)) {}

std::optional<NoiseSimulationParameters> NoiseSimulationParameters::from_xyce_directives(const std::vector<std::string>& directives) {
    // init defaults
    std::string output_node;
    std::string ref_node;
    std::string source_name;
    std::string start_freq_value;
    std::string end_freq_value;
    std::string num_points_value;
    std::string sweep_type;
    std::vector<DeviceNoiseOperator> device_noise_operators;
    std::string data_table_name;
    std::optional<PrintParameters> print_parameters;

    // flag indicating whether a valid directive was found
    bool found = false;

    // parse directives
    for (const auto& directive : directives) {
        // tokenize the directive
        const auto tokens = tokenize(directive);

        // skip empty directives
        if (tokens.empty()) {
            continue;
        }

        const std::string cmd = to_upper(tokens[0]);

        // parse print directives and retain noise-specific output config
        if (cmd == ".PRINT") {
            // parse the print statement from the directive
            const auto print_statement = PrintParameters::from_xyce_statement(directive);
            // retain noise print parameters when found
            if (print_statement) {
                const std::string print_type_upper = to_upper(print_statement->print_type);
                if (print_type_upper == "NOISE") {
                    // extract device noise operators (DNI/DNO) from the output variables
                    std::vector<std::string> filtered_variables;
                    for (const auto& var : print_statement->output_variables) {
                        const std::string var_upper = to_upper(var);
                        if (var_upper.rfind("DNI(", 0) == 0 || var_upper.rfind("DNO(", 0) == 0) {
                            // parse the operator token DNI(node) or DNI(node,source)
                            const size_t open = var.find('(');
                            const size_t close = var.find(')');
                            if (open != std::string::npos && close != std::string::npos && close > open) {
                                const std::string inner = var.substr(open + 1, close - open - 1);
                                const size_t comma = inner.find(',');
                                std::string node = inner;
                                std::string source;
                                if (comma != std::string::npos) {
                                    node = inner.substr(0, comma);
                                    source = inner.substr(comma + 1);
                                }
                                device_noise_operators.emplace_back(var_upper.substr(0, 3), node, source);
                            }
                        }
                        else {
                            filtered_variables.push_back(var);
                        }
                    }
                    // store the parsed print parameters without the operator tokens
                    PrintParameters filtered_print = *print_statement;
                    filtered_print.output_variables = filtered_variables;
                    print_parameters = filtered_print;
                    continue;
                }
            }
        }

        // skip non-NOISE directives
        if (cmd != ".NOISE") {
            continue;
        }

        // flag indicating a valid NOISE directive was found
        found = true;

        // parse output node (position 1); supports V(node), V(node,ref), or a bare node
        if (tokens.size() >= 2) {
            const std::string_view node_token = tokens[1];
            if (to_upper(node_token).substr(0, 2) == "V(" && node_token.size() >= 3 && node_token.back() == ')') {
                // split inner content on comma
                const std::string inner(node_token.substr(2, node_token.size() - 3));
                const size_t comma = inner.find(',');
                if (comma != std::string::npos) {
                    output_node = inner.substr(0, comma);
                    ref_node = inner.substr(comma + 1);
                }
                else {
                    output_node = inner;
                }
            }
            else {
                // bare node
                output_node = node_token;
            }
        }

        // parse source name (position 2)
        if (tokens.size() >= 3) {
            source_name = tokens[2];
        }

        // parse the sweep type and frequency triple starting at position 3
        size_t idx = 3;
        // check for inline DATA table form (DATA=<table>)
        if (idx < tokens.size() && to_upper(tokens[idx]).substr(0, 5) == "DATA=") {
            sweep_type = "DATA";
            data_table_name = tokens[idx].substr(5);
            idx++;
        }
        // check for a leading sweep type keyword
        else if (idx < tokens.size()) {
            const std::string candidate = to_upper(tokens[idx]);
            if (candidate == "LIN" || candidate == "DEC" || candidate == "OCT" || candidate == "DATA") {
                sweep_type = candidate;
                idx++;
                // read the table name for DATA sweeps
                if (candidate == "DATA" && idx < tokens.size()) {
                    data_table_name = tokens[idx];
                    idx++;
                }
            }
        }
        // parse the frequency triple: number of points, start, end
        if (sweep_type != "DATA") {
            if (idx < tokens.size()) {
                num_points_value = tokens[idx];
                idx++;
            }
            if (idx < tokens.size()) {
                start_freq_value = tokens[idx];
                idx++;
            }
            if (idx < tokens.size()) {
                end_freq_value = tokens[idx];
                idx++;
            }
        }
        // default to a LIN sweep when none is specified
        if (sweep_type.empty()) {
            sweep_type = "LIN";
        }
    }

    // return instance if a valid directive was found
    if (!found) {
        return std::nullopt;
    }

    return NoiseSimulationParameters(output_node, ref_node, source_name, start_freq_value, end_freq_value, num_points_value, sweep_type, device_noise_operators, data_table_name, print_parameters);
}

std::vector<std::string> NoiseSimulationParameters::to_xyce_directives(const NetlistTopology& topology) const {
    (void)topology;
    // init output directive list
    std::vector<std::string> directives;
    // build the noise analysis directive
    std::string noise_directive = ".NOISE V(" + output_node;
    // append the reference node
    if (!ref_node.empty()) {
        noise_directive += "," + ref_node;
    }
    noise_directive += ") " + source_name + " ";
    // append the sweep type and frequency triple
    if (!sweep_type.empty() && sweep_type != "DATA") {
        noise_directive += sweep_type + " " + num_points_value + " " + start_freq_value + " " + end_freq_value;
    }
    else {
        // DATA sweep uses the inline table name
        noise_directive += "DATA=" + data_table_name;
    }
    directives.push_back(noise_directive);

    // append the noise print directive with device noise operators
    if (print_parameters && (!print_parameters->output_variables.empty() || !device_noise_operators.empty())) {
        // build the print directive
        std::string print_directive = print_parameters->to_xyce_statement();
        // append the device noise operators
        for (const auto& dno : device_noise_operators) {
            print_directive += " " + dno.type + "(" + dno.node;
            // append the noise source when set
            if (!dno.source.empty()) {
                print_directive += "," + dno.source;
            }
            print_directive += ")";
        }
        directives.push_back(print_directive);
    }

    // return the full directive list
    return directives;
}

bool NoiseSimulationParameters::operator==(const NoiseSimulationParameters& other) const {
    // compare all fields for equality
    return output_node == other.output_node && ref_node == other.ref_node && source_name == other.source_name && start_freq_value == other.start_freq_value && end_freq_value == other.end_freq_value && num_points_value == other.num_points_value && sweep_type == other.sweep_type && device_noise_operators == other.device_noise_operators && data_table_name == other.data_table_name && print_parameters == other.print_parameters;
}
